source/h2_stream.c

```diff
@@ -358,6 +421,67 @@ static void s_stream_update_window(struct aws_http_stream *stream_base, size_t i
     }
 }
 
+static int s_stream_reset_stream(struct aws_http_stream *stream_base, uint32_t http2_error) {
+
+    struct aws_h2_stream *stream = AWS_CONTAINER_OF(stream_base, struct aws_h2_stream, base);
+    struct aws_h2_connection *connection = s_get_h2_connection(stream);
+    bool reset_called;
+    bool stream_is_activating;
+    bool cross_thread_work_should_schedule = false;
+
+    { /* BEGIN CRITICAL SECTION */
+        s_lock_synced_data(stream);
+
+        reset_called = stream->synced_data.reset_called;
+        stream_is_activating = stream->synced_data.api_state == AWS_H2_STREAM_API_STATE_ACTIVATED;
+        if (!reset_called && stream_is_activating) {
+            cross_thread_work_should_schedule = !stream->synced_data.is_cross_thread_work_task_scheduled;
+            stream->synced_data.reset_called = true;
+            stream->synced_data.user_reset_error_code = http2_error;
+        }
+        s_unlock_synced_data(stream);
+    } /* END CRITICAL SECTION */
+
+    if (cross_thread_work_should_schedule) {
+        AWS_H2_STREAM_LOG(TRACE, stream, "Scheduling stream cross-thread work task");
+        /* increment the refcount of stream to keep it alive until the task runs */
+        aws_atomic_fetch_add(&stream->base.refcount, 1);
+        aws_channel_schedule_task_now(connection->base.channel_slot->channel, &stream->cross_thread_work_task);
+        return AWS_OP_SUCCESS;
+    }
+
+    if (!stream_is_activating) {
+        AWS_H2_STREAM_LOG(
+            ERROR, stream, "Reset stream failed. State for the stream doesn't allow RST_STREAM to send. (closed/idle)");
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+
+    if (reset_called) {
+        AWS_H2_STREAM_LOG(ERROR, stream, "Reset stream failed. Reset stream has been called before.");
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
```

Comment:
not sure if this should be an exception ....
like, if the user has multiple codepaths that try to reset the stream if something goes wrong, and multiple things go wrong, might be nicer to just silently ignore the subsequent calls...